text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,266 @@
+
+### Recovering from lifecycle errors
+
+It might happen that an error is thrown while invoking the `connectedCallback` or other lifecycle method. This situation could occur because:
+
+1. The userland lifecycle method has a bug.
+2. The state extracted from the SSR DOM is incorrect for some reason, leading to unexpected inputs and behavior in the lifecycle method.
+
+We will want to recover from #2. And because it won't be possible to differentiate #1 and #2 at runtime, a catch-all solution must be adopted.
+
+It is proposed that, should an error be thrown during `connectedCallback`, we should assume that an unexpected mismatch has occurred, throw away the SSR DOM, and replace it with CSR VDOM. The `connectedCallback` will then be invoked again (`connectedCallback` can fire more than once), this time for the CSR DOM.
+
+If an error occurs again, we can assume its proximate cause is in userland, not LWC.
+
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(MyApp, document.querySelector('#root')); 
+```
+
+### Constructable stylesheets
+
+The introduction of constructable stylesheets to LWC presents a small wrinkle. Since support for this browser API is not yet ubiquitous, we cannot know during SSR whether it will be supported on the client. For that reason, constructable stylesheets are disabled during SSR.
+
+From there, we have two options if support for constructable stylesheets is detected during rehydration:
+
+1. Swap in constructable stylesheets for vanilla stylesheets.
+2. Permanenly disable constructable stylesheets for the lifetime of the `document`.
+
+It is believed that swapping in constructable stylesheets could obviate the performance wins that constructable stylesheets are intended to bring. Furthermore, any theoretical benefit is outweighed by the implementation complexity of swapping in constructable stylesheets.
+
+For these reasons, it is proposed that constructable stylesheets be disabled for the lifetime of the `document` in a post-rehydration context.
+
+### Server-only components
+
+Addressing [Exception B to the happy path](#the-happy-path) will involve a solution roughly comparable to [React server components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html). While the bulk of this can be implemented in user-land, at a level of abstraction above LWC, the `hydrateElement` patch implementation will need to support the use-case.

```

Comment:
What I'm really concerned here is on who has the saying on what parts are rehydratable or not. My intuition tells me that we might want to have some lever at the application level...